### PR TITLE
Add libtorch 

### PR DESCRIPTION
**Pull request by PinkySan**
Specify library name and version:  **lib/1.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

- What should be inside the test_package folder?
- conandata.yml? I am creating the url within the recipe

### CODE REVIEW

File: recipes/libtorch/1.2.0/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        if(self.settings.os == "Windows"):
+            if(self.settings.build_type == "Debug"):
+                return "libtorch-win-shared-with-deps-debug-"
+            return "libtorch-win-shared-with-deps-"
+        if(self.settings.os == "Linux"):
+            if(self.settings.compiler.libcxx == "libstdc++11"):
+                return "libtorch-cxx11-abi-shared-with-deps-"
+            return "libtorch-shared-with-deps-"
+    
+    def build(self):
```
Comment (Croydon) [inline recipes/libtorch/1.2.0/conanfile.py]: There shouldn't be downloads in `build()` please do this in `source()`

Comment (PinkySan) [reply recipes/libtorch/1.2.0/conanfile.py]: renamed the build(self) into source(self)

**Comment by uilianries** [comment]
I've checked this recipe and it's complex ... To build it from source, it requires other libraries. I've tried to create a generic version for any version here: https://github.com/uilianries/conan-center-index/commit/3b007a84431e897f6d62b9c1d21811f0e4511fb9

**Comment by SSE4** [comment]
@uilianries may you create a PR to author's repo with your changes? (+ hook fixes)

**Comment by uilianries** [comment]
@SSE4 Niet, but I should.

**Comment by PinkySan** [comment]
There is a way of installing pytorch from sources 
https://github.com/pytorch/pytorch#from-source

- How to install dependent python packages within conanfile.py?
- Does this include the libtorch packages?

**Comment by danimtb** [comment]
@PinkySan there is no way to install additional dependencies from python. All the resources should be in the recipe.

There are some things missing in this PR, like the test_package recipe and the missing attributes

**Comment by PinkySan** [comment]
> @PinkySan there is no way to install additional dependencies from python. All the resources should be in the recipe.
> 
> There are some things missing in this PR, like the test_package recipe and the missing attributes

Unfortunatly, that is true.
I do not really understand, why they are trying to build the package by their own. Providing the binaries should be enough. (I am having the same difficulties with OpenCV)

**Comment by forwardmeasure** [comment]
@PinkySan, for what it's worth, I have put together a Conan recipe to build Libtorch from scratch. It  works, but not consistently, often failing on this error:

( "You found a build system error: AVX512F, AVX512DQ, AVX512VL " "is defined (via e.g. -mavx512f, -mavx512dq, and -mavx512vl) " "but CAFFE2_PERF_WITH_AVX512 is not defined.")

You can find the recipe here: https://github.com/forwardmeasure/conan/tree/master/recipes/conan-libtorch

Regards,
Prashanth

**Comment by bmanga** [comment]
For what it's worth, I also have a WIP implementation of a libtorch recipe that reuses some packages: https://github.com/bmanga/conan-libtorch

**Comment by forwardmeasure** [comment]
> For what it's worth, I also have a WIP implementation of a libtorch recipe that reuses some packages: https://github.com/bmanga/conan-libtorch

Thanks, Bruno. I got mine working soon after posting. Cheers!

**Comment by danimtb** [comment]
@PinkySan please address the issues reported above by the bot, otherwise the PR will be closed. Thanks!

**Comment by PinkySan** [comment]
I guess this solution works better. Please close this PR
> For what it's worth, I also have a WIP implementation of a libtorch recipe that reuses some packages: https://github.com/bmanga/conan-libtorch

**Comment by bmanga** [comment]
For what it's worth, I'm planning to cleanup my version and propose a barebone version for inclusion soon-ish. I think it would be good to converge and work on a single attempt (there seem to be at least 4 versions based on this PR alone)

**Comment by blackliner** [comment]
@bmanga any updates on this? I did not find another open PR for libtorch, and as we migrate to conan, we might need it soon. Willing to help out 👍

**Comment by SpaceIm** [comment]
I'm working on pytorch recipe, but if someone has a good recipe almost ready for a PR, I'll be happy to help.

**Comment by blackliner** [comment]
I am living with distributing binaries for now :-/
```python
from conans import ConanFile, tools


class LibtorchConan(ConanFile):
    name = "libtorch"
    settings = "os", "arch"
    url = "https://github.com/pytorch/pytorch"
    description = "An open source machine learning framework that accelerates the path from research prototyping to production deployment."
    license = "All rights reserved."
    topics = ("computer-vision", "meta-data")

    def build(self):
        tools.get(**self.conan_data["binaries"][self.version][str(self.settings.arch)])

    def package(self):
        self.copy("*", src=self.name)

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)
```
`conandata.yml`:
```yaml
binaries:
  1.4.0:
    x86_64:
      url: https://download.pytorch.org/libtorch/cu101/libtorch-cxx11-abi-shared-with-deps-1.4.0.zip
      sha256: f214bfde532877aa5d4e0803e51a28fa8edd97b6a44b6615f75a70352b6b542e
```

**Comment by PinkySan** [comment]
@blackliner That is the same idea, i had at the beginning of this PR. I guess that is the best solution

**Comment by blackliner** [comment]
Yes, I guess some things do not meant to be built from scratch :D

I forgot to mention, to use my previous recipe, you need to add `cmake_paths` generator!

**Pull request closed without merging**